[AOTInductor] Memory leak fix for Fallback Kernels

muchulee8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155642

Summary:
We generate AtenTensorHandles for Fallback kernels regardless of the arg
type. If we indeed "fallback", we will regenerate the AtenTensorHandles
that will cause the first handle being generated not recycled, thus a
memory leak would occur.

Test Plan:
python test/inductor/test_aot_inductor.py -k test_fallback_mem_leak

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @amjames @chauhang @aakhundov

muchulee8 committed:
[AOTInductor] Memory leak fix for Fallback Kernels
Summary:
We generate AtenTensorHandles for Fallback kernels regardless of the arg
type. If we indeed "fallback", we will regenerate the AtenTensorHandles
that will cause the first handle being generated not recycled, thus a
memory leak would occur.

Test Plan:
python test/inductor/test_aot_inductor.py -k test_fallback_mem_leak

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jingsh reviewed (approved):
lgtm

desertfire commented on test/inductor/test_aot_inductor.py:
```diff
@@ -2283,6 +2283,57 @@ def forward(self, x):
+        device: int = device_interface.current_device()
+        mem_before = device_interface.memory_allocated(device)
+        aot_inductor_module(*example_inputs)
+        aot_inductor_module(*example_inputs)
```
Is calling twice intentional?

muchulee8 replied:
I changed it to one time, it should be sufficient.

desertfire reviewed (commented):
Can you help me to understand why this is a problem for user-defined Triton kernel in particular?

muchulee8 commented:
> Can you help me to understand why this is a problem for user-defined Triton kernel in particular?

I just extracted a minimal repro from the internal model and replaced the triton kernel with a external one.
I removed the triton kernel and it also show mem-leak, just updated the test.

muchulee8 committed:
Update on "[AOTInductor] Memory leak fix for Fallback Kernels"


Summary:
We generate AtenTensorHandles for Fallback kernels regardless of the arg
type. If we indeed "fallback", we will regenerate the AtenTensorHandles
that will cause the first handle being generated not recycled, thus a
memory leak would occur.

Test Plan:
python test/inductor/test_aot_inductor.py -k test_fallback_mem_leak

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng amjames chauhang aakhundov

[ghstack-poisoned]

muchulee8 commented:
@pytorchbot merge

muchulee8 committed:
Update on "[AOTInductor] Memory leak fix for Fallback Kernels"


Summary:
We generate AtenTensorHandles for Fallback kernels regardless of the arg
type. If we indeed "fallback", we will regenerate the AtenTensorHandles
that will cause the first handle being generated not recycled, thus a
memory leak would occur.

Test Plan:
python test/inductor/test_aot_inductor.py -k test_fallback_mem_leak

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng amjames chauhang aakhundov

[ghstack-poisoned]
